_posts/2019-03-21-timezone-in-rails-and-javascript.md

```diff
@@ -0,0 +1,96 @@
+
+<div class="information information-warning">
+  In almost all scenarios, we don't want to change this value, and we want to keep UTC as the default timezone for Database.
+</div>
+
+If all you're doing is receiving user data through a form and use Active Record to persist it you're good to go. But what about actually doing something with the time information before persisting it? That's when it becomes tricky.
+
+When parsing time information it's important to never do it without specifying the timezone. The best way to do this is to use the methods that use the timezone specified in our rails application config. Let's look at a couple of methods that parses date and time:
+
+**Methods**                           | **Description**
+===================================== | ===============
+`Time.current`                        | uses timezone from rails application config
+`Time.now`                            | uses timezone from system locale
+`Time.zone.xxx`                       | uses timezone from rails application config
+`Time.strptime(xx, xx).in_time_zone`  | uses timezone from rails application config
+`Time.strptime(xx, xx)`               | uses timezone from system locale
+`Date.current`                        | uses timezone from rails application config
+`Date.today`                          | uses timezone from system locale
+`String.to_time`                      | uses timezone from system locale
+`String.in_time_zone`                 | uses timezone from rails application config
+
+Looking carefully, you'll discover that the naming of the methods that adopt rails application configured timezone has a set of common patterns:
+1. we should use the `.current` method whenever available
+2. we should use the `zone` object as intermediate object whenever available
+3. we could use `.in_time_zone` on any value to reset its timezone to rails application config
+
+
+## Timezone in Browser (JS Date API)
+
+The JS Date API is tricky and extra attention is required when creating a Date object from string:
+
+<div class="information information-warning">
+  The Date constructor will always return a DateTime object in browser's local time, but the argument string is taken care of differently when it is provided in different formats.
+</div>
+
+**format**                | **Result**        | **Description**
+==========                | ==========        | ===============
+`new Date('2013-02-22')`  | 'Feb 21 2013 ...' | this treats argument as UTC time
+`new Date('2013/02/22')`  | 'Feb 22 2013 ...' | this treats argument as local time
+`new Date('02/22/2013')`  | 'Feb 22 2013 ...' | this treats argument as local time
+`new Date('22 Feb 2013')` | 'Feb 22 2013 ...' | this treats argument as local time
+
+
+<br>
+## Summary
+
+Timezone issue is hard to discover and tricky to deal with, we should always keep the timezone piece in mind when dealing with date and time.
```

Comment:
"Timezone issues are hard... ...to deal with. We should alway keep timezones in the back of our minds when ..."

Reply:
👍